source/known-hook-events.ts

```diff
@@ -1,8 +1,47 @@
-export default [
+import {RequestOptions} from 'https';
+import {IncomingMessage} from 'http';
+
+// TODO: This may be a bit overkill. It was done for different reasons:
+//       - Expose hook types individually for consumers of the API to easily
+//         type - check their own code that may rely on those specific hooks.
+//
+//        There are some cons about this, though:
+//
+//        - Doesn't properly provide the function signature which ends up hiding type information such as the arguments and return types that could be helpful.
+export type BeforeErrorHook = (error: Error) => void;
+export type InitHook = (options: RequestOptions) => void;
+export type BeforeRequestHook = (options: RequestOptions) => void | Promise<void>
+export type BeforeRedirectHook = (options: RequestOptions) => void | Promise<void>
+export type BeforeRetryHook = (options: RequestOptions) => void | Promise<void>
+// TODO: Update the return type for the retry handler to be an extended PCancelable promise type with extra properties.
+export type AfterResponseHook = (response: IncomingMessage, retryWithMergedOptions: (options: unknown) => any) => void | Promise<void>
+
+export type HookType = BeforeErrorHook | InitHook | BeforeRequestHook | BeforeRedirectHook | BeforeRetryHook | AfterResponseHook;
+
+export interface Hooks {
+	beforeError: BeforeErrorHook[];
+
+	/**
+	 * Called with plain request options, right before their normalization. This is especially useful in conjunction with got.extend() and got.create() when the input needs custom handling.
+	 *
+	 * @default []
+	 */
+	init: InitHook[];
+	beforeRequest: BeforeRequestHook[];
+	beforeRedirect: BeforeRedirectHook[];
+	beforeRetry: BeforeRetryHook[];
+	afterResponse: AfterResponseHook[];
```

Comment:
Should the docs be here or on the actual types? For example, line 11.

Reply:
This was on oversight on my part. I think I used this to check if you could document each property, and I think you can. I’ll suggest we do it on the properties as well as the types, since both can be consumed by different means.

Reply:
Sorry I got it confused with on the type of `Hooks` vs. the properties that takes the list of hooks.

I tried to add the documentation at the type level, but TypeScript/VSCode apparently can't show proper intellisense when you type a function in the array. I think it makes sense to add them to the properties in that case.

Reply:
So the consensus is to add the same documentation for both types and the properties in Hooks?

Reply:
Yup